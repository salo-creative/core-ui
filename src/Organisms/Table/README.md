# Table

The table organism is made up of many different components to deliver a truly flexible and customisable data table. The table supports column sorting, loading states, error handling filtering, pagination and searching

---

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import Table from '@salo/core-ui/Organisms/Table';
```

Implement as follows

```javascript
<Table
  columns={ [
    { label: 'Column 1', dataKey: 'column_1', minWidth: '200px' },
    { label: 'Column 2', dataKey: 'column_2', minWidth: '200px' },
    { label: 'Column 3', dataKey: 'column_3', minWidth: '200px' }
  ] }
  data={ [
    { 'column_1': 'My first thing', 'column_2': 'My second thing', 'column_3': 'My third thing' },
    { 'column_1': 'My first thing', 'column_2': 'My second thing', 'column_3': 'My third thing' },
    { 'column_1': 'My first thing', 'column_2': 'My second thing', 'column_3': 'My third thing' }
  ] }
/>
```

To get a basic table to render you only need to pass the columns and data props.

### Columns

The columns prop expects an array and this tells the table what columns it should have, their minimum width and what key in the data objects to get the data to hydrate the row in the table from. 

*n.b. the minWidth of a column should always be set and should always be a px value. The component will calculate the total of all minWidths and set the mobile breakpoint accordingly*

```javascript
[
	{
		label: 'Column label', // heading that apears in the table head
		dataKey: 'column_key', // where the data for each row is located
		minWidth: '200px', // used to ensure column never gets too small
		width: '50px', // used to ensure column is always this size if it's important the content is a set width
		sortable: false // bool value of whether column should be sortable
	},
	...
]
```

### Data

The data prop expects an array of data objects to be passed in, where each object corresponds to a row in the table. The data that is placed in each column will be determined based on the dataKey value given in the columns array.

```javascript
[ 
	{ 
		'column_key': 'My first thing',
		'column_key_2': 'My second thing', 
		'column_key_3': 'My third thing'
	},
	...
]
```
If the data array is empty then the table will automatically show a user message stating `There are no results to display` so it is clear the table is not broken. This text can be overriden using the `dataEmptyText` prop which simply accepts a string.

## Loading & Error handling

As the table will mostly be used for handling remote resources it has inbuilt loading and error states as well as retry logic.

### Loading

The loader can be toggled via the `loading` prop which accepts a boolean value. The loading state will take precedence over the tables error state.

```javascript
<Table
  columns={ [ ... ] }
  data={ [ ... ] }
  loading={ true }
/>
```

### Error

A simple error message can be toggled via the `error` prop which accepts a boolean value. If this is set to true and the loading state is false then an error message will be displayed within the table. The default error message is `Something went wrong getting your data!` but this can be overriden via the `errorMessage` prop which accepts a string.

```javascript
<Table
  columns={ [ ... ] }
  data={ [ ... ] }
  error={ true }
/>
```

### Retry

In addition to the basic error message, the table also supports the passing of a retry action so failed remote requests can be retried without the user having to refresh. If the `error` prop is true and a `retryAction` has been passed to the table then in addition to the error message outlined above being displayed, a retry button and some expplanation text will also be shown. If the user clicks the button then the function you passeds into the `retryAction` prop will be called.

```javascript
<Table
  columns={ [ ... ] }
  data={ [ ... ] }
  error={ true }
  retryAction={ () => refetchData() } // custom retry function
/>
```

It is important to note the table itself is not aware of what action/request needs to be performed to retry the data fetch, you will need to call this as part of the retry function passed to the table itself.

## Pagination

The table has the pagination molecule bundled with it with some sensible defaults set (for full details on the pagination component see its storybook). To enable the pagination on the table you need to pass both a `pagination` object and a `pageChange` function. e.g.

```javascript
<Table
  columns={ [ ... ] }
  data={ [ ... ] }
  pagination={ {
    perPage: 24,
	  page: 1,
	  total: 240
  } }
  pageChange={ (page) => handlePageChange(page) }
/>
```

the table itself simply displays the pagination controls and exposes a function for handling the page change. It will be the responsibility of your component/view to handle the process of fetching more remote data and passing it to the table to update its contained data and pagination state.

## Actions

The table supports an action area where button etc can be added in the last column of each row. The `action` prop expects a function to be passed which the table then passes the row data to as its only argument. This means you can pass any function containing valid components and have access to the rows information to implement custom actions etc. If a function is passed to the prop then the table will automatically add a new column for the header and each row so there is no need to modify your data or column objects.

```javascript
<Table
  columns={ [ ... ] }
  data={ [ ... ] }
  actions={ (data, context) => (
    <FlyOut context={ context }>
      <FlyOutButton title='Title' onClick={ () => alert(row.column_1) } icon='sync' />
      <FlyOutLink title='Title' link={ `/${ row.column_1 }` } icon='dashboard' />
    </FlyOut>
  ) }
/>
```

For simplicity the `FlyOut` menu molecule has been modified to include a number of specific styles to make it fit with the table perfectly and is the example used above. Because of this the table sizes the action column to 80px to accommodate this but this can be overridden using the `actionsWidth` property which accepts a string px value for the columns width.

*n.b. it is strongly advised for consistency you always use the `FlyOut` molecule as it has been specifically built and modified to work with the table*

## Responsive behaviour

If the parent's width exceeds each column's minimum width (taking into consideration any action buttons on the row) then the table will look like a regular table. If it's lower than each column's minimum width then it'll stack each row to a card layout with the column title above it. We have found that this is a good compromise between scrolling and visibility of data.

## Custom rows

Each element has a number of CSS classes which can be targetted to create custom layouts (e.g. absolutely positioning rows, showing/hiding individual cells, etc).

## Sort (NOT IMPLEMENTED)

**For full prop types and usage see storybook info/knobs**