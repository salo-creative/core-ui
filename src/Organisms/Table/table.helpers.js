import { memoize } from 'lodash';

/* eslint-disable object-curly-newline */
export const determineThreshold = ({ columns, action, actions, actionWidth, actionsWidth }) => {
  // Loop through columns and find minWidths
  let threshold = columns.reduce((accum, column) => {
    if (column.minWidth) {
      return accum + parseInt(column.minWidth, 10);
    }
    return accum;
  }, 0);

  const GUTTER = 20;

  // If an action button is set then account for this.
  if (action) {
    threshold += (parseInt(actionWidth, 10) + GUTTER * 2);
  }
    
  // If actions is set then account for them.
  if (actions) {
    threshold += (parseInt(actionsWidth, 10) + GUTTER * 2);
  }

  return threshold;
};

export const generateRows = memoize((offset = 0) => {
  // Generate 3 rows with 4 columns for use in table and card skeleton layouts.
  // Numbers predetermined but can be dynamically offset to replicate padding.
  return {
    card: [
      [{ x: offset, y: offset + 10 }, { x: offset, y: offset + 55 }, { x: offset, y: offset + 90 }, { x: offset, y: offset + 125 }],
      [{ x: offset, y: offset + 200 }, { x: offset, y: offset + 245 }, { x: offset, y: offset + 280 }, { x: offset, y: offset + 315 }],
      [{ x: offset, y: offset + 400 }, { x: offset, y: offset + 445 }, { x: offset, y: offset + 480 }, { x: offset, y: offset + 515 }]
    ],
    table: [
      [{ x: offset, y: offset }, { x: 220, y: offset + 5 }, { x: 620, y: offset + 5 }, { x: 780, y: offset + 5 }],
      [{ x: offset, y: offset + 55 }, { x: 220, y: offset + 55 + 5 }, { x: 620, y: offset + 55 + 5 }, { x: 780, y: offset + 55 + 5 }],
      [{ x: offset, y: offset + 110 }, { x: 220, y: offset + 110 + 5 }, { x: 620, y: offset + 110 + 5 }, { x: 780, y: offset + 110 + 5 }]
    ]
  };
});