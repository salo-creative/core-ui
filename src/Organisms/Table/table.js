import React from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';

// COMPONENTS & STYLES
import Loader from '../../Molecules/Loader';
import TableProvider from './context/provider';
import TableBody from './table.body';
import TableHeader from './table.header';
import TablePagination from './table.pagination';
import { TableWrapper, LoaderWrapper } from './table.styles';

// HELPERS
import { columnsProps, sortingProps } from './table.propTypes';

const Table = (props) => {
  const {
    action,
    actionWidth,
    actions,
    actionsWidth,
    borders,
    columns,
    className,
    data,
    dataEmptyComponent,
    dataEmptyText,
    error,
    errorMessage,
    loading,
    pager,
    pagination,
    pageChange,
    retryAction,
    rowHeight,
    showHeader,
    sorting,
    onSort,
    width
  } = props;

  const [cardThresholdWidth, setCardThresholdWidth] = React.useState(0);

  const sortMe = ({ dataKey }) => {
    const sortingKey = get(sorting, 'dataKey', null);
    
    // There is no current sorting
    if (sortingKey === null) {
      onSort({
        dataKey,
        direction: 'asc'
      });
    } else if (sortingKey !== dataKey) {
      // Sorting column is being changed
      onSort({
        dataKey,
        direction: 'asc'
      });
    } else {
      // Changing the direction of the sorting column
      const direction = get(sorting, 'direction') === 'asc' ? 'desc' : 'asc';
      onSort({
        dataKey,
        direction
      });
    }
  };

  React.useEffect(() => {
    // Set smallest width before switching to card layout.
    const threshold = columns.reduce((accum, column) => {
      if (column.minWidth) {
        return accum + parseInt(column.minWidth, 10);
      }
      return accum;
    }, 0);

    setCardThresholdWidth(threshold);
  }, [columns]);

  return (
    <TableProvider value={ {
      action,
      actionWidth,
      actions,
      actionsWidth,
      borders,
      columns,
      className,
      data,
      dataEmptyComponent,
      dataEmptyText,
      error,
      errorMessage,
      loading,
      pager,
      pagination,
      pageChange,
      retryAction,
      rowHeight,
      showHeader,
      sorting,
      onSort,
      width,
      sortMe,
      cardThresholdWidth
    } }
    >
      <TableWrapper
        width={ width }
        className={ `salo-table ${ className } ${ borders ? '' : 'no-borders' }` }
      >
        { showHeader && (
          <TableHeader
            hasAction={ !!action }
            hasActions={ !!actions }
          />
        ) }
        { /* Render body if we aren't loading */ }
        { !loading && (
          <TableBody />
        ) }
        { /* Render loader if we are fetching data */ }
        { loading && (
          <LoaderWrapper>
            <Loader
              display={ true }
            />
          </LoaderWrapper>
        ) }
        { !loading && !error && (
          <TablePagination />
        ) }
      </TableWrapper>
    </TableProvider>
  );
};

Table.defaultProps = {
  action: null,
  actionWidth: '12rem',
  actions: null,
  actionsWidth: '8rem',
  borders: true,
  columns: [],
  className: '',
  data: [],
  dataEmptyComponent: null,
  dataEmptyText: 'There are no results to display',
  error: false,
  errorMessage: 'Something went wrong getting your data!',
  loading: false,
  pager: true,
  retryAction: null,
  rowHeight: '6rem',
  showHeader: true,
  sorting: {},
  onSort: () => null,
  width: '100%',
  pagination: null,
  pageChange: null
};

Table.propTypes = {
  action: PropTypes.func,
  actionWidth: PropTypes.string,
  actions: PropTypes.func,
  actionsWidth: PropTypes.string,
  borders: PropTypes.bool,
  columns: columnsProps,
  className: PropTypes.string,
  data: PropTypes.arrayOf(PropTypes.object),
  dataEmptyComponent: PropTypes.any,
  dataEmptyText: PropTypes.string,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  loading: PropTypes.bool,
  pager: PropTypes.bool,
  retryAction: PropTypes.func,
  rowHeight: PropTypes.string,
  showHeader: PropTypes.bool,
  sorting: sortingProps,
  onSort: PropTypes.func,
  width: PropTypes.string,
  pagination: PropTypes.shape({
    perPage: PropTypes.number,
    page: PropTypes.number,
    pages: PropTypes.number,
    total: PropTypes.number
  }),
  pageChange: PropTypes.func
};

export default Table;