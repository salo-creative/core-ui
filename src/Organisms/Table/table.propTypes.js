import PropTypes from 'prop-types';

/**
 * Column props
 */
export const columnProps = PropTypes.shape({
  dataKey: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  minWidth: PropTypes.string.isRequired,
  render: PropTypes.func,
  sortable: PropTypes.bool,
  width: PropTypes.string
});

export const columnsProps = PropTypes.arrayOf(columnProps);

/**
 * Sorting props
 */
export const sortingProps = PropTypes.shape({
  direction: PropTypes.oneOf(['asc', 'desc']),
  dataKey: PropTypes.string
});