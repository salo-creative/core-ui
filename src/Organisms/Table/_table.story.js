import React from 'react';
import { get } from 'lodash';
import { Store, RenderWithProps } from '@jamesbliss/storybook-state';

// load tests
import { withTests } from '@storybook/addon-jest';
import {
  withKnobs, text, object, boolean
} from '@storybook/addon-knobs';
import results from '../../../.storybook/jest-test-results.json';

// FEATURED COMPONENT
import {
  Table, FlyOut, FlyOutLink, FlyOutButton
} from '../../index';

// README //
import README from './README.md';
import Button from '../../Molecules/Button/index.js';

// Start of story logic
export const Basic = () => {
  const margin = text('Width', '100%');
  const width = text('Margin', '0 0 2rem');
  const dataEmptyText = text('Data empty text', 'There are no results to display');
  const borders = boolean('Borders', true);
  const error = boolean('Error', false);
  const errorMessage = text('Error message', 'Something went wrong getting your data!');
  const showHeader = boolean('Show header', true);
  const loading = boolean('Loading', false);
  const columns = object('Columns', [
    {
      label: 'Column 1',
      dataKey: 'column_1',
      minWidth: '200px'
    },
    {
      label: 'Column 2',
      dataKey: 'column_2',
      minWidth: '200px'
    },
    {
      label: 'Column 3',
      dataKey: 'column_3',
      minWidth: '200px'
    }
  ]);
  const data = object('Data', [
    {
      column_1: 'My first thing',
      column_2: 'My second thing',
      column_3: 'My third thing'
    },
    {
      column_1: 'My first thing',
      column_2: 'My second thing',
      column_3: 'My third thing'
    },
    {
      column_1: 'My first thing',
      column_2: 'My second thing',
      column_3: 'My third thing'
    }
  ]);
  return (
    <Table
      borders={ borders }
      columns={ columns }
      data={ data }
      dataEmptyText={ dataEmptyText }
      error={ error }
      errorMessage={ errorMessage }
      loading={ loading }
      margin={ margin }
      retryAction={ () => alert('retry') }
      showHeader={ showHeader }
      width={ width }
    />
  );
};

Basic.story = {
  decorators: [
    withKnobs,
    withTests({
      results
    })
  ],
  parameters: {
    notes: README,
    jest: ['table']
  }
};

export const Pagination = () => {
  const margin = text('Width', '100%');
  const width = text('Margin', '0 0 2rem');
  const dataEmptyText = text('Data empty text', 'There are no results to display');
  const borders = boolean('Borders', true);
  const error = boolean('Error', false);
  const errorMessage = text('Error message', 'Something went wrong getting your data!');
  const showHeader = boolean('Show header', true);
  const loading = boolean('Loading', false);
  const pager = boolean('Pager', true);
  const pagination = object('Pagination', {
    perPage: 24,
    page: 3,
    pages: 13,
    total: 300
  });
  const columns = object('Columns', [
    {
      label: 'Column 1',
      dataKey: 'column_1',
      minWidth: '200px'
    },
    {
      label: 'Column 2',
      dataKey: 'column_2',
      minWidth: '200px'
    },
    {
      label: 'Column 3',
      dataKey: 'column_3',
      minWidth: '200px'
    }
  ]);
  const data = object('Data', [
    {
      column_1: 'My first thing',
      column_2: 'My second thing',
      column_3: 'My third thing'
    },
    {
      column_1: 'My first thing',
      column_2: 'My second thing',
      column_3: 'My third thing'
    },
    {
      column_1: 'My first thing',
      column_2: 'My second thing',
      column_3: 'My third thing'
    },
    {
      column_1: 'My first thing',
      column_2: 'My second thing',
      column_3: 'My third thing'
    },
    {
      column_1: 'My first thing',
      column_2: 'My second thing',
      column_3: 'My third thing'
    },
    {
      column_1: 'My first thing',
      column_2: 'My second thing',
      column_3: 'My third thing'
    }
  ]);
  return (
    <Table
      borders={ borders }
      columns={ columns }
      data={ data }
      dataEmptyText={ dataEmptyText }
      error={ error }
      errorMessage={ errorMessage }
      loading={ loading }
      margin={ margin }
      pager={ pager }
      pagination={ pagination }
      pageChange={ page => alert(`Change to page => ${ page }`) }
      retryAction={ () => alert('retry') }
      showHeader={ showHeader }
      width={ width }
    />
  );
};

Pagination.story = {
  decorators: [
    withKnobs,
    withTests({
      results
    })
  ],
  parameters: {
    notes: README,
    jest: ['table']
  }
};

export const Actions = () => {
  const margin = text('Width', '100%');
  const width = text('Margin', '0 0 2rem');
  const dataEmptyText = text('Data empty text', 'There are no results to display');
  const borders = boolean('Borders', true);
  const error = boolean('Error', false);
  const errorMessage = text('Error message', 'Something went wrong getting your data!');
  const loading = boolean('Loading', false);
  const showHeader = boolean('Show header', true);
  const showSingleAction = boolean('Show single action', false);
  const showActions = boolean('Show action menu', true);
  const columns = object('Columns', [
    {
      label: 'Column 1',
      dataKey: 'column_1',
      minWidth: '200px'
    },
    {
      label: 'Column 2',
      dataKey: 'column_2',
      minWidth: '200px'
    },
    {
      label: 'Column 3',
      dataKey: 'column_3',
      minWidth: '200px'
    }
  ]);
  const data = object('Data', [
    {
      column_1: 'My first thing',
      column_2: 'My second thing',
      column_3: 'My third thing'
    },
    {
      column_1: 'My first thing',
      column_2: 'My second thing',
      column_3: 'My third thing'
    },
    {
      column_1: 'My first thing',
      column_2: 'My second thing',
      column_3: 'My third thing'
    }
  ]);
  const inlineMenu = (row, layout) => (
    <FlyOut context={ layout }>
      <FlyOutButton title='Title' onClick={ () => alert(row.column_1) } icon='sync' />
      <FlyOutLink title='Title' link={ `/${ row.column_1 }` } icon='dashboard' />
    </FlyOut>
  );
  const action = row => (
    <Button onClick={ () => alert(row.column_1) } iconAfter='sync'>
      Title
    </Button>
  );
  return (
    <Table
      borders={ borders }
      columns={ columns }
      data={ data }
      dataEmptyText={ dataEmptyText }
      error={ error }
      errorMessage={ errorMessage }
      loading={ loading }
      margin={ margin }
      showHeader={ showHeader }
      width={ width }
      { ...(showActions && {
        actions: inlineMenu
      }) }
      { ...(showSingleAction && {
        action
      }) }
    />
  );
};

Actions.story = {
  decorators: [
    withKnobs,
    withTests({
      results
    })
  ],
  parameters: {
    notes: README,
    jest: ['table']
  }
};

// STATE
const store = new Store({
  sorting: {
    direction: 'desc',
    dataKey: 'column_1'
  },
  data: [
    {
      column_1: '1',
      column_2: '4',
      column_3: '7'
    },
    {
      column_1: '2',
      column_2: '5',
      column_3: '8'
    },
    {
      column_1: '3',
      column_2: '6',
      column_3: '9'
    }
  ]
});

export const Sorting = () => {
  const margin = text('Width', '100%');
  const width = text('Margin', '0 0 2rem');
  const dataEmptyText = text('Data empty text', 'There are no results to display');
  const borders = boolean('Borders', true);
  const error = boolean('Error', false);
  const errorMessage = text('Error message', 'Something went wrong getting your data!');
  const loading = boolean('Loading', false);
  const showHeader = boolean('Show header', true);
  const columns = object('Columns', [
    {
      label: 'Column 1',
      dataKey: 'column_1',
      minWidth: '200px',
      sortable: true
    },
    {
      label: 'Column 2',
      dataKey: 'column_2',
      minWidth: '200px',
      sortable: true
    },
    {
      label: 'Column 3',
      dataKey: 'column_3',
      minWidth: '200px'
    }
  ]);

  return (
    <RenderWithProps store={ store }>
      <Table
        borders={ borders }
        columns={ columns }
        sorting={ store.state.storing }
        onSort={ sorting => {
          const { dataKey, direction } = sorting;
          const data = get(store, 'state.data', []);

          const sortedData = data.sort((a, b) => {
            const comparison = get(a, dataKey, '').localeCompare(get(b, dataKey, ''));
            return direction === 'desc' ? -comparison : comparison;
          });

          console.log(sortedData);

          store.set({
            sorting,
            data
          });
        } }
        data={ store.state.data }
        dataEmptyText={ dataEmptyText }
        error={ error }
        errorMessage={ errorMessage }
        loading={ loading }
        margin={ margin }
        showHeader={ showHeader }
        width={ width }
      />
    </RenderWithProps>
  );
};

Sorting.story = {
  decorators: [
    withKnobs,
    withTests({
      results
    })
  ],
  parameters: {
    jest: ['table'],
    info: {
      propTablesExclude: [RenderWithProps]
    },
    notes: README
  }
};

export default {
  title: 'Organisms/Table'
};