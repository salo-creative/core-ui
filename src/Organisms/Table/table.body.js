import React from 'react';
import PropTypes from 'prop-types';
import { isEmpty } from 'lodash';

// COMPONENTS & STYLES
import Row from './table.row';
import { BodyWrapper, NoData, ErrorMessage } from './table.styles';
import P from '../../Typography/P';
import Button from '../../Molecules/Button';

// HELPERS
import { columnsProps } from './table.propTypes';

const Body = (props) => {
  const {
    actions,
    actionsWidth,
    columns,
    data,
    dataEmptyComponent,
    dataEmptyText,
    error,
    errorMessage,
    retryAction
  } = props;

  // If an error let the user know and show a retry action
  if (error) {
    // If we have retry action display message and button
    if (retryAction) {
      return (
        <ErrorMessage>
          <P align='center'><strong>{ errorMessage }</strong></P>
          <P align='center'>You can retry the last action by pressing the button below.</P>
          <Button onClick={ retryAction } iconBefore='sync'>Retry</Button>
        </ErrorMessage>
      );
    }
    // Otherwise just show the message
    return (
      <NoData>{ errorMessage }</NoData>
    );
  }
  // If no data display a friendly message for the user
  if (isEmpty(data)) {
    if (dataEmptyComponent) return dataEmptyComponent;
    return (
      <NoData>{ dataEmptyText }</NoData>
    );
  }
  return (
    <BodyWrapper>
      { data.map((item, i) => {
        const key = `${ item[Object.keys(item)[0]] }-${ i }`;
        return (
          <Row
            key={ key }
            actions={ actions }
            actionsWidth={ actionsWidth }
            columns={ columns }
            data={ item }
          />
        );
      }) }
    </BodyWrapper>
  );
};

Body.defaultProps = {
  actions: null,
  columns: [],
  data: [],
  dataEmptyComponent: null,
  retryAction: null
};

Body.propTypes = {
  actions: PropTypes.any,
  actionsWidth: PropTypes.string.isRequired,
  columns: columnsProps,
  data: PropTypes.arrayOf(PropTypes.object),
  dataEmptyComponent: PropTypes.any,
  dataEmptyText: PropTypes.string.isRequired,
  error: PropTypes.bool.isRequired,
  errorMessage: PropTypes.string.isRequired,
  retryAction: PropTypes.func
};

export default Body;