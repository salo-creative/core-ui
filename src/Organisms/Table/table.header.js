import React from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';

// COMPONENTS & STYLES
import Icon from '@salo/icons';
import { HeaderRow, HeaderCell, HeaderSorting } from './table.styles';

// HELPERS
import { columnsProps, sortingProps } from './table.propTypes';
import { colours } from '../../helpers/colours';


const Header = (props) => {
  const { actionsWidth, columns, hasActions, sorting, onSort } = props;

  // Determines the sorting icon to be rendered
  const sortingIcon = (dataKey) => {
    const sortingKey = get(sorting, 'dataKey', null);

    if (sortingKey === null || sortingKey !== dataKey) {
      return 'chevron_up_down';
    }

    return get(sorting, 'direction') === 'asc' ? 'chevron_up' : 'chevron_down';
  };

  return (
    <HeaderRow>
      { columns.map(column => {
        const { label, minWidth, dataKey, sortable } = column;
        const icon = sortingIcon(dataKey);
        return (
          <HeaderCell
            key={ dataKey }
            flexBasis={ `${ 100 / columns.length }%` }
            minWidth={ minWidth }
          >
            { label }
            { sortable && (
              <HeaderSorting onClick={ () => onSort({ dataKey }) }>
                <Icon size={ 24 } fill={ icon === 'chevron_up_down' ? colours.grey : colours.black } icon={ icon } />
              </HeaderSorting>
            ) }
          </HeaderCell>
        );
      }) }
      { hasActions && (
        <HeaderCell
          key='actions'
          flexBasis={ actionsWidth }
          minWidth={ actionsWidth }
        />
      ) }
    </HeaderRow>
  );
};

Header.defaultProps = {
  columns: [],
  sorting: {},
  onSort: () => null
};

Header.propTypes = {
  actionsWidth: PropTypes.string.isRequired,
  columns: columnsProps,
  sorting: sortingProps,
  onSort: PropTypes.func,
  hasActions: PropTypes.bool.isRequired
};

export default Header;