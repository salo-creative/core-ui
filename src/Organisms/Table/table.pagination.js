import React from 'react';
import PropTypes from 'prop-types';
import { isEmpty } from 'lodash';

// COMPONENTS & STYLES
import Pagination from '../../Molecules/Pagination';
import Button from '../../Molecules/Button';
import TableContext from './context/context';
import { LoadMoreWrapper } from './table.styles';

const TablePagination = () => {
  const {
    loading,
    pager,
    pagination,
    pageChange
  } = React.useContext(TableContext);

  // Bailout if we have no pagination data or functions
  if (isEmpty(pagination) || !pageChange) return null;

  // Return pager style nav
  if (pager) {
    return (
      <Pagination
        { ...pagination }
        pagesToShow={ 7 }
        changePage={ pageChange }
      />
    );
  }

  // Render load more if we have pages of data left
  if (pagination.pages > pagination.page) {
    return (
      <LoadMoreWrapper>
        <Button
          className='salo-table__pagination-button'
          loading={ loading }
          onClick={ () => pageChange(pagination.page + 1) }
        >
            Load more
        </Button>
      </LoadMoreWrapper>
    );
  }

  return null;
};

TablePagination.defaultProps = {
  loading: false,
  pager: true,
  pagination: null,
  pageChange: null
};

TablePagination.propTypes = {
  loading: PropTypes.bool,
  pager: PropTypes.bool,
  pagination: PropTypes.shape({
    perPage: PropTypes.number.isRequired,
    page: PropTypes.number.isRequired,
    pages: PropTypes.number.isRequired,
    total: PropTypes.number
  }),
  pageChange: PropTypes.func
};

export default TablePagination;