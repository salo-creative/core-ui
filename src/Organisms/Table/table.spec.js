import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import Table from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <Table />
  );
  expect(container.firstChild).toMatchSnapshot();
});