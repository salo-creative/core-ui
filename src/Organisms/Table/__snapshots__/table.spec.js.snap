// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c0 {
  background-image: linear-gradient( 90deg,rgba(255,255,255,0) 0,rgba(255,255,255,.8) 50%,rgba(255,255,255,0) 100% ),linear-gradient(#E1E1E1 25px,transparent 0),linear-gradient(#E1E1E1 15px,transparent 0),linear-gradient(#E1E1E1 15px,transparent 0),linear-gradient(#E1E1E1 15px,transparent 0),linear-gradient(#E1E1E1 25px,transparent 0),linear-gradient(#E1E1E1 15px,transparent 0),linear-gradient(#E1E1E1 15px,transparent 0),linear-gradient(#E1E1E1 15px,transparent 0),linear-gradient(#E1E1E1 25px,transparent 0),linear-gradient(#E1E1E1 15px,transparent 0),linear-gradient(#E1E1E1 15px,transparent 0),linear-gradient(#E1E1E1 15px,transparent 0),linear-gradient(white 100%,transparent 0);
  background-size: 200px 100%,120px 25px,250px 15px,100px 15px,140px 15px,120px 25px,250px 15px,100px 15px,140px 15px,120px 25px,250px 15px,100px 15px,140px 15px,100% 100%;
  background-position: -150% 0,0px 0px,220px 5px,620px 5px,780px 5px,0px 55px,220px 60px,620px 60px,780px 60px,0px 110px,220px 115px,620px 115px,780px 115px,0 0;
  background-repeat: no-repeat;
  -webkit-animation: iRoVfX 2.5s infinite;
  animation: iRoVfX 2.5s infinite;
  width: 100%;
}

.c0 .salo-table__header-row,
.c0 .salo-table__body-wrapper {
  visibility: hidden;
}

.c1 {
  display: none;
  width: 100%;
  height: 5rem;
  -webkit-flex-wrap: none;
  -ms-flex-wrap: none;
  flex-wrap: none;
  border-bottom: 2px solid #e9ecef;
  position: relative;
}

.no-borders .c1 {
  border-bottom: none;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1.6rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2.error {
  color: #f15b60;
}

@media only screen and (max-width:0px) {
  .c0 {
    -webkit-animation: iZOQfl 2.5s infinite;
    animation: iZOQfl 2.5s infinite;
    background-position: -150% 0,0px 0px,0px 45px,0px 80px,0px 115px,0px 190px,0px 235px,0px 270px,0px 305px,0px 390px,0px 435px,0px 470px,0px 505px,0 0;
  }
}

<div
  class="c0 salo-table  "
  width="100%"
>
  <div
    class="c1 salo-table__header salo-table__header-row"
  />
  <div
    class="c2 salo-table__no-data"
  >
    There are no results to display
  </div>
</div>
`;
